### PR TITLE
Add `package_cache_async` flag.

### LINKED ISSUE #1379: Should be possible to configure that a resolve blocks until caching completes (and use the cache for all items)
Subject says most of it.

Ideally this would be implemented as a rezconfig flag, and perhaps rez-env or similar would have a flag to ignore the caching step if a user merely wishes to check the resolve. `--solve-only`?

I imagine that this sort of workflow would be a requirement of s3-bucket-style repositories anyway, so we ought to support this in general.

**Motivation**
Some users may wish to enforce usage of only cached packages, and simply wait for caching to complete.

### PR DESCRIPTION
**Pull request by ttrently**
Fixes #1379.

Adds a `package_cachy_async` flag which allows users to run caching synchronously (blocking) or asynchronously from the config.

This setting is set to `True` by default in the rezconfig so that behavior should act as normal. Setting this to false will block the resolve until all packages have been cached successfully.

### CODE REVIEW

**Comment by JeanChristopheMorinPerso** [comment]
Thanks @ttrently ! Can you follow the instructions at https://github.com/AcademySoftwareFoundation/rez/pull/1452/checks?check_run_id=11526898738 to fix the DCO check please?

**Comment by alx00x** [comment]
A welcome addition. I want to add it should be implemented as a flag on `rez-env` command and an argument for `ResolvedContext` in the same manner how package caching itself can be disabled or enabled explicitly (config being a fallback).

For example:

```
rez env my_package --no-async
```

```
context = ResolvedContext(["my_package"], async_package_caching=False)
```

File: src/rez/rezconfig.py [hunk 1]
```diff
@@ -269,6 +269,9 @@
 # Enable package caching during a package build.
 package_cache_during_build = False
 
+# Enable package caching to run asynchronously during a resolve.
```
Comment (JeanChristopheMorinPerso) [inline src/rez/rezconfig.py:272]: I find this hard to read. How about "set to false to always cache package payloads synchronously"?

**Review by JeanChristopheMorinPerso — CHANGES_REQUESTED** [review]
Thanks @ttrently ! I think it could be a good idea to have a CLI flag, but I'm not sure what it should be named and to which commands it should apply. Any opinions?

**Comment by JeanChristopheMorinPerso** [comment]
Also, the linter is complaining: https://github.com/AcademySoftwareFoundation/rez/actions/runs/4245498949/jobs/7387520293#step:5:1

File: src/rez/package_cache.py [hunk 2]
```diff
@@ -465,8 +465,14 @@ def add_variants_async(self, variants):
                 else:
                     out_target = devnull
 
-                subprocess.Popen(
-                    [exe, "--daemon", self.path],
```
[the remark this answers is not in the document]
Comment (ttrently) [reply src/rez/package_cache.py:469]: These are defined on line 457 but the `args` variable was never used unless I missed something. The `func` call on line 474 now uses that instead.

[the remark this answers is not in the document]
Comment (JeanChristopheMorinPerso) [reply src/rez/package_cache.py:469]: My bad. I got confused by how the diff looks like.

**Comment by JeanChristopheMorinPerso** [comment]
Loses in favor of #1679. @ttrently I'll make sure that you are marked as co-author.

**Pull request closed without merging**